service/history/workflowcache/cache.go

```diff
@@ -133,16 +133,29 @@ func (c *wfCache) allow(domainID string, workflowID string, rateLimitType rateLi
 
 	switch rateLimitType {
 	case external:
-		return value.externalRateLimiter.Allow()
+		if !value.externalRateLimiter.Allow() {
+			c.emitMetrics(domainID, workflowID, domainName)
+			return false
+		}
+		return true
 	case internal:
-		return value.internalRateLimiter.Allow()
+		if !value.internalRateLimiter.Allow() {
+			c.emitMetrics(domainID, workflowID, domainName)
+			return false
+		}
+		return true
 	default:
 		// This should never happen, and we fail open
 		c.logError(domainID, workflowID, errors.New("unknown rate limit type"))
 		return true
 	}
 }
 
+func (c *wfCache) emitMetrics(domainID string, workflowID string, domainName string) {
```

Comment:
tbh I think I'd name this more `emitLimited...` or similar.  "emit metrics" is completely unclear that it's only for one purpose, and elsewhere we use that kind of name for bulk-flushing aggregate metrics which isn't happening here.